src/saml.py

```diff
@@ -66,6 +84,31 @@ def metadata_url(self) -> str:
         """
         return self._charm_state.metadata_url
 
+    @cached_property
+    def _is_certificate_valid(self):
+        """Validate the public certificate for the metadata URL.
+
+        Returns:
+            True if the certificate matches the one provided.
+
+        Raises:
+            TimeoutError: if the connection can't be established.
+        """
+        if self._charm_state.fingerprint:
+            url = urllib.parse.urlparse(self._charm_state.metadata_url)
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.settimeout(1)
+            context = ssl.create_default_context()
+            with socket.create_connection((url.hostname, 443)) as sock:
+                with context.wrap_socket(sock, server_hostname=url.hostname) as wrapped_socket:
```

Comment:
Just noticed, minor nitpick,
I think these with statements can be merged to a single with statement.